src/libraries/Microsoft.PowerFx.Core/Utils/TextFormatUtils.cs

```diff
@@ -85,21 +97,154 @@ public static bool IsValidFormatArg(string formatString, CultureInfo formatCultu
+                }
+                else if (formatStr[i] == ':')
+                {
+                    hasColonWithNum = true;
+                }
+                else if (formatStr[i] == ';')
+                {
+                    // Does not allow number of section are more than 2.
+                    sectionCount++;
+                    if (sectionCount > 2)
+                    {
+                        return false;
+                    }
+                }
+                else if (i == formatStr.Length - 1)
+                {
+                    // If format string ends with backsplash but no following character (both numeric and datetime format), format is invalid.
+                    if (formatStr[i] == '\\')
+                    {
+                        return false;
+                    }
+
+                    // If format string ends with e or e+ (not escaping character) then format is invalid (only applied for numeric format)
+                    if (textFormatArgs.HasNumericFmt & (formatStr[i] == 'e' || (i > 2 && formatStr[i - 2] != '\\' && formatStr[i - 1] == 'e' && formatStr[i] == '+')))
+                    {
+                        return false;
+                    }
+                }
+                else if (formatStr[i] == '\\')
+                {
+                    if (textFormatArgs.HasNumericFmt || (i < formatStr.Length - 1 && formatStr[i + 1] == '\"'))
+                    {
+                        // Skip next character if seeing escaping character \.
+                        i++;
+                    }
+                    else
+                    {
+                        // Update \c to "c" to match with                       
+                        formatStr = formatStr.Insert(i + 2, "\"");
+                        formatStr = formatStr.Insert(i + 1, "\"");
+                        formatStr = formatStr.Remove(i, 1);
+                        i += 2;
+                    }
+                }
+                else if (formatStr[i] == '\"' && i < formatStr.Length - 1)
+                {
+                    // Jump to close quote to pass all escaping characters.
+                    i = formatStr.IndexOf('\"', i + 1);
+
+                    // Format is invalid if missing close quote.
+                    if (i == -1)
+                    {
+                        return false;
+                    }
+                }
+
+                if (textFormatArgs.HasDateTimeFmt && textFormatArgs.HasNumericFmt)
+                {
+                    // Check if the date time format contains '0's after the seconds specifier, which
+                    // is used for fractional seconds - in which case it is valid
+                    var formatWithoutZeroSubseconds = _formatWithoutZeroSubsecondsRegex.Replace(formatStr, m => m.Groups[1].Success ? string.Empty : m.Groups[1].Value);
+                    textFormatArgs.HasNumericFmt = formatWithoutZeroSubseconds.IndexOfAny(_numericCharacters.ToArray()) >= 0;
```

Comment:
>_numericCharacters.ToArray() [](http://example.com/codeflow?start=90&length=28)

No need to call `ToArray()` on every iteration

Reply:
i will move this out of the loop